File: backend/src/api/api_schemas/journal/metrics.py
import datetime as dt
import uuid
from typing import Optional
from pydantic import BaseModel, Field


class MetricSchema(BaseModel):
    """Response schema for metric data"""

    id: uuid.UUID
    thread_id: uuid.UUID
    asleep_by: dt.datetime | None
    awoke_at: dt.datetime | None
    out_of_bed_at: dt.datetime | None
    sleep_quality: int | None
    physical_activity: int | None
    overall_mood: int | None
    paid_productivity: float | None
    personal_productivity: float | None
    additional_metrics: dict | None
    created_at: dt.datetime
    updated_at: dt.datetime

    class Config:

        from_attributes = True


class MetricCreateSchema(BaseModel):
    """Schema for creating a new metric"""

    thread_id: uuid.UUID
    asleep_by: Optional[dt.datetime] = None
    awoke_at: Optional[dt.datetime] = None
    out_of_bed_at: Optional[dt.datetime] = None
    sleep_quality: Optional[int] = Field(None, ge=1, le=7)
    physical_activity: Optional[int] = Field(None, ge=1, le=7)
    overall_mood: Optional[int] = Field(None, ge=1, le=7)
    paid_productivity: Optional[float] = None
    personal_productivity: Optional[float] = None
    additional_metrics: Optional[dict] = None


class MetricUpdateSchema(BaseModel):
    """Schema for full metric update"""

    thread_id: uuid.UUID
    asleep_by: Optional[dt.datetime] = None
    awoke_at: Optional[dt.datetime] = None
    out_of_bed_at: Optional[dt.datetime] = None
    sleep_quality: Optional[int] = Field(None, ge=1, le=7)
    physical_activity: Optional[int] = Field(None, ge=1, le=7)
    overall_mood: Optional[int] = Field(None, ge=1, le=7)
    paid_productivity: Optional[float] = None
    personal_productivity: Optional[float] = None
    additional_metrics: Optional[dict] = None


class MetricPatchSchema(BaseModel):
    """Schema for partial metric update"""

    id: uuid.UUID
    thread_id: Optional[uuid.UUID] = None
    asleep_by: Optional[dt.datetime] = None
    awoke_at: Optional[dt.datetime] = None
    out_of_bed_at: Optional[dt.datetime] = None
    sleep_quality: Optional[int] = Field(None, ge=1, le=7)
    physical_activity: Optional[int] = Field(None, ge=1, le=7)
    overall_mood: Optional[int] = Field(None, ge=1, le=7)
    paid_productivity: Optional[float] = None
    personal_productivity: Optional[float] = None
    additional_metrics: Optional[dict] = None


class MetricUpsertSchema(BaseModel):
    """Schema for upsert operations - checks for existing record by thread_id"""

    thread_id: uuid.UUID
    asleep_by: Optional[dt.datetime] = None
    awoke_at: Optional[dt.datetime] = None
    out_of_bed_at: Optional[dt.datetime] = None
    sleep_quality: Optional[int] = Field(None, ge=1, le=7)
    physical_activity: Optional[int] = Field(None, ge=1, le=7)
    overall_mood: Optional[int] = Field(None, ge=1, le=7)
    paid_productivity: Optional[float] = None
    personal_productivity: Optional[float] = None
    additional_metrics: Optional[dict] = None

File: frontend/src/types/metrics.ts
// Backend metric schema matching
export interface Metric {
  id: string;
  thread_id: string;
  asleep_by: string | null; // ISO datetime string
  awoke_at: string | null; // ISO datetime string
  out_of_bed_at: string | null; // ISO datetime string
  sleep_quality: number | null;
  physical_activity: number | null; // minutes or intensity score
  overall_mood: number | null;
  paid_productivity: number | null;
  personal_productivity: number | null;
  additional_metrics: Record<string, unknown> | null;
  created_at: string; // ISO datetime string
  updated_at: string; // ISO datetime string
}

// Frontend-friendly metric structure for a specific date
export interface DailyMetrics {
  date: string; // ISO date string (YYYY-MM-DD)
  thread_id?: string;
  asleep_by?: string | null; // ISO datetime string
  awoke_at?: string | null; // ISO datetime string
  out_of_bed_at?: string | null; // ISO datetime string
  sleep_quality?: number | null; // 1-7 scale
  physical_activity?: number | null; // 1-7 scale
  overall_mood?: number | null; // 1-7 scale
  paid_productivity?: number | null;
  personal_productivity?: number | null;
}

export interface MetricsSummary {
  from: string; // ISO date string
  to: string; // ISO date string
  averages: {
    sleep_quality?: number;
    physical_activity?: number;
    overall_mood?: number;
    paid_productivity?: number;
    personal_productivity?: number;
    [key: string]: number | undefined;
  };
  data: DailyMetrics[];
}

